File: tests/simple_kernel.py
import sys
from trusimd import *

# Expect one argument
if len(sys.argv) == 1:
    print('{}: error: usage: {} search_string'. \
          format(sys.argv[0], sys.argv[0]))
    sys.exit(1)

# Poll hardware and select hardware based on argv[1]
hardwares = poll_hardware()
h = [x for x in hardwares \
     if x.description.lower().find(sys.argv[1].lower()) >= 0]

if len(h) == 0:
    print('{}: info: no hardware could be selected'.format(sys.argv[0]))
else:
    h = h[0]
    print('{}: info: selected {}'.format(sys.argv[0], h))

# Create memory buffers
n = 1024
a = buffer_pair(h, n, float32)
b = buffer_pair(h, n, float32)
c = buffer_pair(h, n, float32)

# Fill buffers with numbers
for i in range(n):
    b[i] = i
    c[i] = i

# Kernel
with kernel('vector_add', float32ptr, float32ptr, float32ptr) as vector_add:
    arg(0)[gid] = arg(1)[gid] + arg(2)[gid]

# Print Kernel source code for debugging
print(vector_add)

# Copy data to device, compile and execute kernel
b.copy_to_device()
c.copy_to_device()
vector_add.run(h, n, a, b, c);

# Check result
a.copy_to_host();
for i in range(n):
    if a[i] != b[i] + c[i]:
        print('{}: error: {} vs. {}'.format(sys.argv[0], a[i], b[i] + c[i]))
        sys.exit(-1)
    else:
        print('{}: info: {} --> {}'.format(sys.argv[0], a[i], b[i] + c[i]))

File: tests/simple_kernel.cpp
#include <trusimd.hpp>
#include <iostream>
#include <cstring>

int main(int argc, char **argv) {
  using namespace trusimd;

  // Expect one argument
  if (argc != 2) {
    std::cerr << argv[0] << ": error: usage: " << argv[0]
              << " search_string\n";
    return -1;
  }

  // Poll hardware and select hardware based on argv[1]
  hardware &h = find_hardware(argv[1]);
  std::cout << argv[0] << ": info: selected " << h.description << '\n';

  // Create memory buffers
  const int n = 4096;
  buffer_pair<float> a(h, n), b(h, n), c(h, n);

  // Fill buffers with numbers
  for (int i = 0; i < n; i++) {
    b[i] = float(i);
    c[i] = float(i);
  }

  // Kernel
  kernel vector_add("vector_add", float32ptr, float32ptr, float32ptr);
  {
    arg(0)[gid] = arg(1)[gid] + arg(2)[gid];
  }

  // Print Kernel source code for debugging
  //std::cout << vector_add << std::endl;

  // Copy data to device, compile and execute kernel
  b.copy_to_device();
  c.copy_to_device();
  vector_add(h, n, a, b, c);

  // Check result
  a.copy_to_host();
  for (int i = 0; i < n; i++) {
    if (a[i] != b[i] + c[i]) {
      std::cerr << argv[0] << ": error: " << a[i] << " vs. " << (b[i] + c[i])
                << std::endl;
      return -1;
    } else {
      std::cout << argv[0] << ": info: " << a[i] << " --> " << (b[i] + c[i])
                << std::endl;
    }
  }

  return 0;
}
